### ISSUE TITLE
Support formatting to and from deep JSON object

### ISSUE THREAD

**Issue by etsuo** [sakuraapi/api#126]
#124 #96 were implemented on 0.9.0, but fail to take into account some scenarios.

## The Problem

There are times when incoming JSON from a service may not be in the the shape we want. By shape, I mean that it's not simply field name transformation that we want to accomplish, it's also morphing the shape of the incoming JSON such that it results in a Model that has a different shape.

Incoming JSON:
```
const data = {
    "first_name": "john",
    "source": {
        "address_1": "123 main st",
        "address_2": "suite 100
    }
}
```
Desired result from `SomeModel.fromJson(data, 'source1')`:
```
{
  firstName: 'john',
  address: '123 main st suite 100"
}
```
It may also be the case that the incoming JSON comes in multiple shapes depending on the source, but the resulting Model is common to those various sources. For example the following might also be incoming JSON from a different source that would result in `SomeModel` above.
```
const data = {
    "firstName": "john",
    "address": "123 main st suite 100"
}
```

And it should be the case that you can marshall a Model back to its original shape (`someModel.toJson('source1')`):
```
{
    "first_name": "john",
    "source": {
        "address_1": "123 main st",
        "address_2": "suite 100
    }
}
```
or with `someModel.toJson()`:
```
const data = {
    "firstName": "john",
    "address": "123 main st suite 100"
}
```

### Approach 1: Field Level Trasformation
```
@Model()
class SomeModel {
  @Json()
  @Json('first_name', 'source1')
  firstName: string;

  @Json()
  @Json({
    field: 'source',
    formatFromJson: this.buildModelAddress,
    formatToJson: this.buildJsonAddress,
  }, 'source1')
  address: string;

  private buildModelAddress(val, key) {
    return `${val.address_1} ${val.address_2}`;
  }

  private buildJsonAddress(val, key, obj) {
    const parts = val.split(' ');
    obj.source = obj.source || {};
    obj.source.address_1 = parts[0];
    obj.source.address_2 = parts.length > 1 ? parts[0] : undefined;
    return val;
  }
}
```

For `formatFromJson`, `val` will be the child object `source` (which `fromJson` will not recurse into).

For `formatToJson`, `val` will be the address string of the model, and a new parameter, `obj` will be passed in, which is the root of the object being built.

### Approach 2: Model Level Transformation
It might be nice to also have model level formatters that get applied at the end of `toJson` or `fromJson`:
```
@Model()
class SomeModel {
  @Json()
  @Json('first_name', 'source1')
  firstName: string;

  @Json()
  address: string;

  @FormatFromJson('source1')
  private buildModelAddress(json, model, context) {
    if (context !== 'source1') return result;

    result.address  = `${val.address_1} ${val.address_2}`;
    return result;
  }

  @FormatToJson('source1')
  private buildJsonAddress(model, json, context) {
    if (context !== 'source1') return result;

    const parts = (source.address || '').split(' ');
    result.source = result.source || {};
    result.source.address_1 = parts[0];
    result.source.address_2 = parts.length > 1 ? parts[0] : undefined;
    return result;
  }
}
```
The former approach makes sense if you have one off fields that need custom formatting. The latter makes sense of you're going to be transforming a bunch of stuff.

**Comment by candleshine** [comment]
I think another use case we are running into for something like this is our image fields.  It might be good to strip off the repetitive parts of the file names, and be able to programmatically add prefix/suffixes that are needed.  

Example, incoming format: 
```
{
  imageUrl: 'http://mycdn.com/10-Africa.hires.png'
}
```
outgoing format:
```
{
  thumbnailUrl: 'http://mycdn.com/10-Africa.thumbnail.png',
  lowResImageUrl: 'http://mycdn.com/10-Africa.lowres.png',
  hiResImageUrl: 'http://mycdn.com/10-Africa.hires.png'
}
```
Maybe.

**Comment by etsuo** [comment]
@candleshine this should be doable with Approach 2 above.

**Comment by etsuo** [comment]
Approach 2 was released with 0.10.0. Approach was is tracked in #128.
